### ISSUE TITLE
SqlException within concurrent call to ExecuteReader returns bad connections to the pool 

### ISSUE THREAD

**Issue by burns47** [dotnet/SqlClient#1887]
Under certain conditions, it's possible for a call to ExecuteReader to return results and/or errors from another instance of SqlCommand.  For context, we're seeing this bug when calling Dapper's Query method.  The following conditions must be met for this bug to occur:

1.  A connection is being used in an unsafe way
2. The two threads must be calling `ExecuteReader`, and at least one of the commands must throw a SqlException
3. A third, unrelated thread, using its own isolated `SqlConnection`, should run its own query at roughly the same time

Under these conditions, the third thread mentioned above may receive the error or result set from one of the first two threads' queries.

The bug occurs much more often if `ExecuteReader(CommandBehavior.CloseConnection)` is used.

The bug does _not_ occur if:

1. Neither of the two threads throws an error
2. `ExecuteReaderAsync` is used instead

### To reproduce

See attached zip solution/project
[ConnectionPoolPoisoning.zip](https://github.com/dotnet/SqlClient/files/10411851/ConnectionPoolPoisoning.zip)

### Expected behavior
Errors coming from thread-unsafe usage of SqlConnection shouldn't cause errors in other, well-behaving parts of the application.

### Further technical details
Microsoft.Data.SqlClient version: 5.0.1
.NET target: .NET 6.0
SQL Server version: SQL Server Express, Azure SQL Managed Instance as well
Operating system: Windows 10

I was also able to reproduce this with System.Data.SqlClient.

**Comment by lcheunglci** [comment]
Thanks @burns47 for bringing this issue to our attention and providing a repro sample.  We're currently in the middle of a release.  We'll get back to you soon.

**Comment by JRahnama** [comment]
@burns47 I have not looked into the details of this issue, but in general Connections are not thread safe. We have had several issues  such as #1758.

**Comment by burns47** [comment]
I agree that we shouldn't expect to be able to use a connection from multiple threads - however I was hoping that the resulting undefined behavior would be isolated to the connection that you were using in an unsafe way.  That doesn't seem to be the case.  If you're using a SqlConnection in a thread-unsafe manner, errors can appear in _any part_ of the application using SqlConnections, even if they're using their own connection safely.  I was hoping there would be a way to limit the blast radius of such a bug!

**Comment by Wraith2** [comment]
There is. Don't ever use connections in an unsafe and unsupported way and you won't have a problem.

**Comment by roji** [comment]
FWIW I do agree that it's important for buggy user code to not cause completely undefined/unpredictable behavior, as much as possible; such behavior makes the bug that much harder to track down. Developers do occasionally introduce bugs in code, and it's important for components to take that into account and be as friendly/safe as possible.

**Comment by lcheunglci** [comment]
Hi @burns47 , this seems like a bug, we'll investigate.

**Comment by cheenamalhotra** [comment]
👀 (sneaking in..) 

I have been looking into it, tried capturing traces and it seems to be related to the 'stateObj' re-assignment (you can run the repro in debug and notice a debug assert failing..), which seems to be causing exceptions/results left unread and flow into the pool from where next connection picks up. 

I agree, this should be fixed to avoid any surprises for unintentional designs.

**Comment by lcheunglci** [comment]
Thanks @cheenamalhotra for your investigation. We'll look into a fix after the release.
